lib/hooks/userspace/hook-grpc.js

```diff
@@ -129,6 +128,299 @@ function makeClientConstructorWrap(makeClientConstructor) {
   };
 }
 
+// # Server
+
+/**
+ * A helper function to record metadata in a trace span. The return value of this
+ * function can be used as the 'wrapper' argument to wrap sendMetadata.
+ * sendMetadata is a member of each of ServerUnaryCall, ServerWriteableStream,
+ * ServerReadableStream, and ServerDuplexStream.
+ * @param rootContext The span object to which the metadata should be added.
+ * @returns {Function} A function that returns a wrapped form of sendMetadata.
+ */
+function sendMetadataWrapper(rootContext) {
+  return function (sendMetadata) {
+    return function sendMetadataTrace(responseMetadata) {
+      if (rootContext) {
+        rootContext.addLabel('metadata', JSON.stringify(responseMetadata.getMap()));
+      } else {
+        agent.logger.info('gRPC: No root context found in sendMetadata');
+      }
+      return sendMetadata.apply(this, arguments);
+    };
+  };
+}
+
+/**
+ * Wraps a bidirectional streaming function in order to record trace spans.
+ * @param {cls.Namespace} namespace The CLS namespace.
+ * @param {Object} handlerSet An object containing references to the function handle,
+ * as well as serialize and deserialize handles.
+ * @param {string} requestName The human-friendly name of the request.
+ */
+function wrapBidi(namespace, handlerSet, requestName) {
+  var rootContext;
+  // CLS context object
+  var context = {};
+  // Start right before call to function
+  // End right after output stream 'finish' event
+  shimmer.wrap(handlerSet, 'func', function (func) {
+    return namespace.bind(function funcTrace(stream) {
+      rootContext = startRootSpanForRequest(requestName);
+      if (agent.config_.enhancedDatabaseReporting) {
+        shimmer.wrap(stream, 'sendMetadata', sendMetadataWrapper(rootContext));
+      }
+      // Preserve context in stream event handlers.
+      namespace.bindEmitter(stream);
+      var spanEnded = false;
+      var endSpan = function () {
+        if (!spanEnded) {
+          spanEnded = true;
+          endRootSpanForRequest(rootContext, requestName);
+        }
+      };
+      stream.on('finish', function () {
+        // End the span unless there is an error.
+        if (stream.status.code === 0) {
+          endSpan();
+        }
+      });
+      stream.on('error', function (err) {
+        if (agent.config_.enhancedDatabaseReporting) {
+          rootContext.addLabel('error', err);
```

Comment:
From what I've seen on the GCP Trace UI it displays the error in plain text (as opposed to JSON). It should be the same string as on the client side. I believe the tests confirm this.